apollo-normalized-cache-incubating/src/appleMain/kotlin/com/apollographql/apollo3/cache/normalized/internal/Guard.kt

```diff
@@ -0,0 +1,31 @@
+package com.apollographql.apollo3.cache.normalized.internal
+
+import kotlinx.coroutines.runBlocking
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+
+internal actual class Guard<R : Any> actual constructor(name: String, producer: () -> R) {
+  private val mutex = Mutex()
+  private val resource = producer()
+
+  actual suspend fun <T> readAccess(block: (R) -> T): T {
+    return mutex.withLock {
+      block(resource)
+    }
+  }
+
+  actual suspend fun <T> writeAccess(block: (R) -> T): T {
+    return mutex.withLock {
+      block(resource)
+    }
+  }
+
+  actual fun writeAndForget(block: (R) -> Unit) = runBlocking {
+    mutex.withLock {
+      block(resource)
+    }
+  }
+
+  actual fun dispose() {
+  }
+}
```

Comment:
Looks like all the `Guard` implementations are the same now? Can you simplify everything and put the mutex in `DefaultApolloStore` ?

Reply:
Did this in 1de5b7592c6b805ce4742dfd49f713f1e9cdb4e3.

On the JVM, we used to use a `ReentrantReadWriteLock`, which has a different behavior for read and write - I'm wondering if we could lose some performance by using a simple Mutex instead.

Reply:
LGTM. Still needs to be done for `-incubating` artifacts